src/tests/selection/test_selection.fypp

```diff
@@ -0,0 +1,369 @@
+              call check( (kth_smallest == i**2), " ${name}$: kth smallest entry with right specified")
+          end do
+
+          ! Simple tests
+          mat = one * [3, 2, 7, 4, 5, 1, 4, -1]
+          mat_copy = mat
+          call select(mat_copy, 1_ip, kth_smallest)
+          call check( kth_smallest == -1, " ${name}$: mat test 1")
+          mat_copy = mat
+          call select(mat_copy, 2_ip, kth_smallest)
+          call check( kth_smallest == 1, " ${name}$: mat test 2")
+          mat_copy = mat
+          call select(mat_copy, size(mat, kind=ip)+1_ip-4_ip, kth_smallest)
+          call check( kth_smallest == 4, " ${name}$: mat test 3")
+          mat_copy = mat
+          call select(mat_copy, 5_ip, kth_smallest)
+          call check( kth_smallest == 4, " ${name}$: mat test 4")
+          mat_copy = mat
+          call select(mat_copy, 6_ip, kth_smallest)
+          call check( kth_smallest == 4, " ${name}$: mat test 5")
+          mat_copy = mat
+          call select(mat_copy, 7_ip, kth_smallest)
+          call check( kth_smallest == 5, " ${name}$: mat test 6")
+
+          ! Check it works for size(a) == 1
+          len1(1) = -1 * one
+          call select(len1, 1_ip, kth_smallest)
+          call check(kth_smallest == -1, " ${name}$: array with size 1")
+
+          ! Check it works for size(a) == 2
+          len2 = [-3, -5]*one
+          call select(len2, 2_ip, kth_smallest)
+          call check(kth_smallest == -3, " ${name}$: array with size 2, test 1")
+          len2 = [-3, -5]*one
+          call select(len2, 1_ip, kth_smallest)
+          call check(kth_smallest == -5, " ${name}$: array with size 2, test 2")
+          len2 = [-1, -1]*one
+          call select(len2, 1_ip, kth_smallest)
+          call check(kth_smallest == -1, " ${name}$: array with size 2, test 3")
+          len2 = [-1, -1]*one
+          call select(len2, 2_ip, kth_smallest)
+          call check(kth_smallest == -1, " ${name}$: array with size 2, test 4")
+
+          !
+          ! Test using random data
+          !
+          any_failed=.FALSE.
+
+          ! Search for the p-th smallest rank, for all these p
+          ! (avoid end-points to enable constrained search tests)
+          do p = 3, Nr-2
+
+              ! Repeat for different random samples to try to expose any errors
+              do i = 1, Nreps
+
+                  ! Make random numbers of the correct type
+                  call random_number(random_doubles)
+                  random_vals = random_doubles * Nr
+
+                  call select(random_vals, p, kth_smallest)
+
+                  test1 = kth_smallest == random_vals(p)
+                  test2 = all(random_vals(1:(p-1)) <= random_vals(p))
+                  test3 = all(random_vals(p) <= &
+                      random_vals((p+1):size(random_vals, kind=ip)))
+                  if( (.not. test1) .or. (.not. test2) .or. (.not. test3) ) &
+                      any_failed = .TRUE.
```

Comment:
Can this be checks? Carrying around a logical will obfuscate the final report from check and make the tests harder to debug.

Reply:
We could do that, although the double loop means that we would then have about 250 checks of each of `test1, test2 and test3`. Are there any downsides to that? If it fails I presume it will make a lot of noise (which might well be OK...?).  

Subjectively I'm not sure if it would be any easier to debug. If any of these tests fails, it means the random numbers have exposed some kind of "corner case" that was not properly treated by the algorithm. To debug it, likely one would have to throw a lot of reproducible random numbers at the code until we found one that failed -- and work backward from that.  

So while I don't mind making the changes, I would like to double-check that there are not downsides from your view?

Reply:
@awvwgk I adjusted the code so that testdrive's `check` is used in each of these cases (see [this commit](https://github.com/fortran-lang/stdlib/pull/500/commits/573dbfadadf8f7c4ffb2070e8e6579058628ff13)). 

So we are no longer carrying around a logical variable to cover all the random tests. 

There do not seem to be any performance issues (at least in the sense that the tests were always very fast, and remain so).